File: 20-actors/uchiwake/methods/crosscheck.py
#!/usr/bin/env python3
"""uchiwake 内訳 — kabuto coverage-linkage crosscheck (stdlib only). ADR-2606081800.

uchiwake's product graph references companies (brand-owner, BOM supplier, process
operator, logistics carrier, ownership parent/child) by kabuto :company/id in the
shared org.corp.* space. This tool computes — does not claim — how much of that
product graph actually WIRES INTO kabuto's ingested company universe, and surfaces
the gap honestly (a reference that does not resolve = "not yet ingested in kabuto",
NOT "does not exist"; G5).

It also reports the OWNERSHIP-ROLLUP effect: a brand-owner subsidiary that is NOT
itself in kabuto but whose ULTIMATE parent IS — i.e. the 子会社 edge recovers a link
that the flat reference would have missed.

This directly answers the standing coverage question ("how integrated is the
supply chain across actors?") with a measured percentage, not an assertion.

stdlib only. Usage:
    python3 crosscheck.py            # human report
    python3 crosscheck.py --json     # machine summary
"""
from __future__ import annotations
import sys
import os
import json
import pathlib
from collections import defaultdict

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from uchiwake_edn import load_edn, classify  # noqa: E402

HERE = pathlib.Path(__file__).resolve().parent
ROOT = HERE.parent
SEED = ROOT / "data" / "seed-products.kotoba.edn"
# kabuto seed lives at 20-actors/kabuto/data/seed-public-companies.kotoba.edn
KABUTO_SEED = ROOT.parent / "kabuto" / "data" / "seed-public-companies.kotoba.edn"


def load_kabuto():
    """Return (company_ids:set, supply_out_degree:dict) or (None, None) if absent."""
    if not KABUTO_SEED.is_file():
        return None, None
    ids = set()
    out_degree = defaultdict(int)
    for r in load_edn(KABUTO_SEED):
        if not isinstance(r, dict):
            continue
        if ':company/id' in r:
            ids.add(r[':company/id'])
        elif ':supply.edge/from' in r:  # supplier node = out-edge source
            out_degree[r[':supply.edge/from']] += 1
    return ids, out_degree


def load_kabuto_company_ids():
    ids, _ = load_kabuto()
    return ids


def uchiwake_covered_companies(g):
    """Set of kabuto company ids that have ANY product-level detail in uchiwake."""
    covered = set()
    for p in g['products'].values():
        if p.get(':product/brand-owner'):
            covered.add(p[':product/brand-owner'])
    for e in g['bom']:
        if e.get(':bom.edge/supplier'):
            covered.add(e[':bom.edge/supplier'])
    for s in g['process']:
        if s.get(':process.step/operator'):
            covered.add(s[':process.step/operator'])
    for lg in g['logistics']:
        if lg.get(':logistics.leg/carrier'):
            covered.add(lg[':logistics.leg/carrier'])
    return covered


def collect_company_refs(g):
    """Return {kind: [(ref_id, holder_id), ...]} for every company reference in the graph."""
    refs = defaultdict(list)
    for pid, p in g['products'].items():
        bo = p.get(':product/brand-owner')
        if bo:
            refs['brand-owner'].append((bo, pid))
    for e in g['bom']:
        s = e.get(':bom.edge/supplier')
        if s:
            refs['bom-supplier'].append((s, e[':bom.edge/id']))
    for s in g['process']:
        op = s.get(':process.step/operator')
        if op:
            refs['process-operator'].append((op, s[':process.step/id']))
    for lg in g['logistics']:
        c = lg.get(':logistics.leg/carrier')
        if c:
            refs['logistics-carrier'].append((c, lg[':logistics.leg/id']))
    for o in g['ownership']:
        refs['ownership-child'].append((o[':company.ownership/child'], o[':company.ownership/id']))
        refs['ownership-parent'].append((o[':company.ownership/parent'], o[':company.ownership/id']))
    return refs


def crosscheck():
    g = classify(load_edn(SEED))
    kabuto_ids = load_kabuto_company_ids()
    refs = collect_company_refs(g)
    ownership_index = {o[':company.ownership/child']: o[':company.ownership/parent']
                       for o in g['ownership']}

    def ultimate(cid, _d=0):
        if cid is None or _d > 16:
            return cid
        nxt = ownership_index.get(cid)
        return cid if (nxt is None or nxt == cid) else ultimate(nxt, _d + 1)

    summary = {'kabuto_available': kabuto_ids is not None,
               'kabuto_company_count': len(kabuto_ids) if kabuto_ids else 0,
               'by_kind': {}, 'rollup_recovered': []}

    all_refs = set()
    resolved = set()
    for kind, items in refs.items():
        k_total, k_res = 0, 0
        for ref_id, holder in items:
            all_refs.add(ref_id)
            k_total += 1
            if kabuto_ids and ref_id in kabuto_ids:
                k_res += 1
                resolved.add(ref_id)
            elif kabuto_ids:
                # rollup: subsidiary not in kabuto but ultimate parent is
                up = ultimate(ref_id)
                if up != ref_id and up in kabuto_ids:
                    summary['rollup_recovered'].append({'ref': ref_id, 'ultimate': up, 'kind': kind})
        summary['by_kind'][kind] = {'total': k_total, 'resolved': k_res}

    distinct = sorted(all_refs)
    summary['distinct_company_refs'] = len(distinct)
    summary['distinct_resolved'] = len(resolved)
    summary['linkage_pct'] = round(100.0 * len(resolved) / max(1, len(distinct)), 1)
    summary['unresolved'] = sorted(all_refs - resolved)

    # ── REVERSE coverage: what fraction of kabuto's SUPPLY-CHAIN companies have any
    #    product-level BOM detail in uchiwake? + prioritized worklist of the highest-
    #    centrality kabuto suppliers still missing product detail (the ingest worklist).
    _, out_degree = load_kabuto()
    if kabuto_ids is not None:
        covered = uchiwake_covered_companies(g) & kabuto_ids
        supply_companies = set(out_degree.keys()) if out_degree else set()
        covered_supply = covered & supply_companies
        summary['reverse'] = {
            'kabuto_supply_companies': len(supply_companies),
            'with_product_detail': len(covered_supply),
            'reverse_pct': round(100.0 * len(covered_supply) / max(1, len(supply_companies)), 3),
            'all_company_coverage_pct': round(100.0 * len(covered) / max(1, len(kabuto_ids)), 3),
            # worklist: top kabuto suppliers (by out-degree) with NO uchiwake product detail
            'worklist': [
                {'company': c, 'supply_out_degree': d}
                for c, d in sorted(out_degree.items(), key=lambda kv: -kv[1])
                if c not in covered
            ][:15],
        }
    return summary


def render(s):
    out = ["# uchiwake ⇄ kabuto coverage-linkage crosscheck\n",
           "> Measured (not claimed) integration of the uchiwake product graph into kabuto's",
           "> ingested company universe. Unresolved = \"not yet ingested in kabuto\", not \"nonexistent\" (G5).\n"]
    if not s['kabuto_available']:
        out.append("kabuto seed not found — cannot crosscheck. (expected at 20-actors/kabuto/data/)")
        return "\n".join(out) + "\n"
    out.append(f"- kabuto ingested companies: **{s['kabuto_company_count']}**")
    out.append(f"- distinct company refs in uchiwake: **{s['distinct_company_refs']}**")
    out.append(f"- resolved into kabuto: **{s['distinct_resolved']}** "
               f"(**{s['linkage_pct']}%** linkage)\n")
    out.append("| reference kind | total | resolved |")
    out.append("|---|---:|---:|")
    for kind, v in sorted(s['by_kind'].items()):
        out.append(f"| {kind} | {v['total']} | {v['resolved']} |")
    if s['rollup_recovered']:
        out.append("\n## 子会社 rollup recovered (subsidiary not in kabuto, but ultimate parent is)\n")
        for r in s['rollup_recovered']:
            out.append(f"- `{r['ref']}` → ultimate `{r['ultimate']}` ({r['kind']})")
    if s['unresolved']:
        out.append("\n## Not yet ingested in kabuto (honest gap)\n")
        for u in s['unresolved']:
            out.append(f"- `{u}`")
    rev = s.get('reverse')
    if rev:
        out.append("\n## Reverse coverage — how much of kabuto has product-level BOM detail (情報取得割合)\n")
        out.append(f"- kabuto supply-chain companies (appear as a supplier): **{rev['kabuto_supply_companies']}**")
        out.append(f"- of those, with ANY uchiwake product detail: **{rev['with_product_detail']}** "
                   f"(**{rev['reverse_pct']}%**)")
        out.append(f"- across ALL {s['kabuto_company_count']} kabuto companies: "
                   f"**{rev['all_company_coverage_pct']}%** have product detail")
        out.append("\nThis is the honest worldwide-coverage figure: the product-BOM layer covers a")
        out.append("tiny fraction of the company universe today. Full ingest is R1 / G7-gated.\n")
        if rev['worklist']:
            out.append("### Ingest worklist — highest-centrality kabuto suppliers with NO product BOM yet\n")
            out.append("| kabuto supplier | supply out-degree |")
            out.append("|---|---:|")
            for w in rev['worklist']:
                out.append(f"| `{w['company']}` | {w['supply_out_degree']} |")
    return "\n".join(out) + "\n"


def main(argv):
    s = crosscheck()
    if '--json' in argv:
        print(json.dumps(s, indent=2, ensure_ascii=False))
    else:
        print(render(s))
    return 0


if __name__ == '__main__':
    raise SystemExit(main(sys.argv))

File: 20-actors/uchiwake/methods/crosscheck.clj
;; crosscheck.clj — uchiwake 内訳 ⇄ kabuto coverage-linkage crosscheck.
;;
;; Clojure port of crosscheck.py (ADR-2606081800), Wave 2 of the clj-native migration
;; (ADR-2606142300) — the first uchiwake method ported beyond the already-`.cljc` uchiwake-edn /
;; analyze. Reuses the canonical `uchiwake.methods.uchiwake-edn` (load-edn + classify); this is a
;; clj-side data tool (reads disk seeds), so it is `.clj` rather than `.cljc`.
;;
;; uchiwake's product graph references companies by kabuto :company/id in the shared org.corp.*
;; space. This MEASURES — does not claim — how much of that graph wires into kabuto's ingested
;; company universe, and surfaces the gap honestly (an unresolved ref = "not yet ingested in
;; kabuto", NOT "does not exist"; G5). It also reports the 子会社 OWNERSHIP-ROLLUP effect (a
;; brand-owner subsidiary not itself in kabuto but whose ULTIMATE parent is) and the REVERSE
;; coverage (what fraction of kabuto's supply-chain companies have product-level BOM detail) +
;; a prioritized ingest worklist. stdlib + uchiwake-edn only.
(ns uchiwake.methods.crosscheck
  (:require [uchiwake.methods.uchiwake-edn :as ue]
            [clojure.set :as set]
            [clojure.java.io :as io])
  (:import [java.math BigDecimal RoundingMode]))

(def default-seed "20-actors/uchiwake/data/seed-products.kotoba.edn")
(def default-kabuto-seed "20-actors/kabuto/data/seed-public-companies.kotoba.edn")

(defn- round-n [x scale]
  (.doubleValue (.setScale (BigDecimal/valueOf (double x)) (int scale) RoundingMode/HALF_EVEN)))

(defn load-kabuto
  "Return [company-ids out-degree] or [nil nil] if the kabuto seed is absent. A :company/id row is
   a company; a :supply.edge/from row contributes an out-edge (supplier centrality)."
  [path]
  (if-not (.isFile (io/file path))
    [nil nil]
    (reduce (fn [[ids od] r]
              (cond
                (not (map? r))                  [ids od]
                (contains? r ":company/id")     [(conj ids (get r ":company/id")) od]
                (contains? r ":supply.edge/from") [ids (update od (get r ":supply.edge/from") (fnil inc 0))]
                :else                           [ids od]))
            [#{} {}]
            (ue/load-edn path))))

(defn uchiwake-covered-companies
  "Set of kabuto company ids that have ANY product-level detail in uchiwake."
  [g]
  (into #{}
        (concat (keep #(get % ":product/brand-owner") (vals (:products g)))
                (keep #(get % ":bom.edge/supplier") (:bom g))
                (keep #(get % ":process.step/operator") (:process g))
                (keep #(get % ":logistics.leg/carrier") (:logistics g)))))

(defn collect-company-refs
  "Ordered [[kind [[ref-id holder-id] …]] …] for every company reference in the graph (kind order
   mirrors crosscheck.py: brand-owner → bom-supplier → process-operator → logistics-carrier →
   ownership-child → ownership-parent)."
  [g]
  [["brand-owner"       (vec (keep (fn [[pid p]] (when-let [bo (get p ":product/brand-owner")] [bo pid])) (:products g)))]
   ["bom-supplier"      (vec (keep (fn [e] (when-let [s (get e ":bom.edge/supplier")] [s (get e ":bom.edge/id")])) (:bom g)))]
   ["process-operator"  (vec (keep (fn [s] (when-let [op (get s ":process.step/operator")] [op (get s ":process.step/id")])) (:process g)))]
   ["logistics-carrier" (vec (keep (fn [l] (when-let [c (get l ":logistics.leg/carrier")] [c (get l ":logistics.leg/id")])) (:logistics g)))]
   ["ownership-child"   (vec (map (fn [o] [(get o ":company.ownership/child") (get o ":company.ownership/id")]) (:ownership g)))]
   ["ownership-parent"  (vec (map (fn [o] [(get o ":company.ownership/parent") (get o ":company.ownership/id")]) (:ownership g)))]])

(defn- ultimate
  "Follow ownership child→parent edges to the topmost parent (cycle/depth guarded)."
  [own-index cid]
  (loop [c cid d 0]
    (let [nxt (get own-index c)]
      (if (or (nil? nxt) (= nxt c) (> d 16)) c (recur nxt (inc d))))))

(defn crosscheck
  "Measure uchiwake⇄kabuto linkage + reverse coverage. Returns the summary map."
  ([] (crosscheck default-seed default-kabuto-seed))
  ([seed-path kabuto-path]
   (let [g          (ue/classify (ue/load-edn seed-path))
         [kabuto-ids out-degree] (load-kabuto kabuto-path)
         refs       (collect-company-refs g)
         own-index  (into {} (map (fn [o] [(get o ":company.ownership/child") (get o ":company.ownership/parent")]) (:ownership g)))
         all-refs   (into #{} (mapcat (fn [[_ items]] (map first items)) refs))
         resolved   (if kabuto-ids (set/intersection all-refs kabuto-ids) #{})
         by-kind    (into {} (for [[kind items] refs]
                               [kind {:total (count items)
                                      :resolved (count (filter (fn [[r _]] (and kabuto-ids (kabuto-ids r))) items))}]))
         rollup     (vec (for [[kind items] refs
                               [ref-id _]   items
                               :when        (and kabuto-ids (not (kabuto-ids ref-id)))
                               :let         [up (ultimate own-index ref-id)]
                               :when        (and (not= up ref-id) (kabuto-ids up))]
                           {:ref ref-id :ultimate up :kind kind}))
         base       {:kabuto-available (some? kabuto-ids)
                     :kabuto-company-count (if kabuto-ids (count kabuto-ids) 0)
                     :by-kind by-kind
                     :rollup-recovered rollup
                     :distinct-company-refs (count all-refs)
                     :distinct-resolved (count resolved)
                     :linkage-pct (round-n (/ (* 100.0 (count resolved)) (max 1 (count all-refs))) 1)
                     :unresolved (vec (sort (set/difference all-refs resolved)))}]
     (if (nil? kabuto-ids)
       base
       (let [covered        (set/intersection (uchiwake-covered-companies g) kabuto-ids)
             supply-cos     (set (keys out-degree))
             covered-supply (set/intersection covered supply-cos)]
         (assoc base :reverse
                {:kabuto-supply-companies (count supply-cos)
                 :with-product-detail (count covered-supply)
                 :reverse-pct (round-n (/ (* 100.0 (count covered-supply)) (max 1 (count supply-cos))) 3)
                 :all-company-coverage-pct (round-n (/ (* 100.0 (count covered)) (max 1 (count kabuto-ids))) 3)
                 :worklist (->> out-degree
                                (sort-by val >)
                                (remove (fn [[c _]] (covered c)))
                                (take 15)
                                (mapv (fn [[c d]] {:company c :supply-out-degree d})))}))))))
